joerd/download.py

```diff
@@ -0,0 +1,169 @@
+from contextlib import contextmanager, closing
+import urllib2
+import tempfile
+import os
+import logging
+import shutil
+import httplib
+import ftplib
+from time import sleep
+
+
+# Custom error wrapper for (known) exceptions thrown by the download module.
+class DownloadFailedError(Exception):
+    pass
+
+
+@contextmanager
+def get(url, options={}):
+    """
+    Download a file to a temporary directory, returning it.
+
+    The options provided will control the behaviour of the download algorithm.
+
+      * 'tries' - The maximum number of tries to download the file before
+        giving up and raising an exception.
+      * 'timeout' - Timeout in seconds before considering the connection to
+        have failed.
+      * 'verifier' - A function which is called with a filelike object. It
+        should return True if the file is okay and appears to be fully
+        downloaded.
+    """
+    logger = logging.getLogger('download')
+
+    with closing(tempfile.NamedTemporaryFile()) as tmp:
+        # current file position = number of bytes read
+        filepos = 0
+
+        # file size when downloaded, if known
+        filesize = None
+
+        # number of attempts so far
+        tries = 0
+
+        # maximum number of attempts to make
+        max_tries = options.get('tries', 1)
+
+        # timeout for blocking operations (e.g: connect) in seconds
+        timeout = options.get('timeout', 60)
+
+        # verifier function
+        verifier = options.get('verifier')
+
+        # backoff function - to delay between retries
+        backoff = options.get('backoff')
+
+        # whether the server supports Range headers (if it doesn't we'll have
+        # to restart from the beginning every time).
+        accept_range = False
+
+        # we need to download _something_ if the file position is less than the
+        # known size, or the size is unknown.
+        while filesize is None or filepos < filesize:
+            # explode if we've exceeded the number of allowed attempts
+            if tries >= max_tries:
+                raise DownloadFailedError("Max tries exceeded (%d) while "
+                                          "downloading file %r"
+                                          % (max_tries, url))
+            else:
+                if backoff:
+                    backoff(tries)
```

Comment:
Yup - well spotted. I hadn't noticed because the "zeroth" backoff was zero in the exponential scheme I'm using, but it wouldn't necessarily always be that way. I've added some new code in 217ce3e to use the number of retries since the last forward progress as the number passed into the backoff. This ensures that the "zeroth" backoff is always zero.